{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Introduction to Pygenprop\n",
    "=========================\n",
    "An python library for interactive programmatic usage of Genome Properties\n",
    "------------------------------------------------------------------------\n",
    "\n",
    "InterProScan files used in this tutorial can be found at:\n",
    "- https://raw.githubusercontent.com/Micromeda/pygenprop/master/docs/source/_static/tutorial/E_coli_K12.tsv\n",
    "- https://raw.githubusercontent.com/Micromeda/pygenprop/master/docs/source/_static/tutorial/E_coli_K12.faa\n",
    "- https://raw.githubusercontent.com/Micromeda/pygenprop/master/docs/source/_static/tutorial/E_coli_O157_H7.tsv\n",
    "- https://raw.githubusercontent.com/Micromeda/pygenprop/master/docs/source/_static/tutorial/E_coli_O157_H7.faa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creation and use of GenomePropertyTree objects\n",
    "GenomePropertyTree objects allow for the programmatic exploration of the Genome properties database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from io import StringIO\n",
    "from pygenprop.results import GenomePropertiesResults, GenomePropertiesResultsWithMatches, \\\n",
    "    load_assignment_caches_from_database, load_assignment_caches_from_database_with_matches\n",
    "from pygenprop.database_file_parser import parse_genome_properties_flat_file\n",
    "from pygenprop.assignment_file_parser import parse_interproscan_file, \\\n",
    "    parse_interproscan_file_and_fasta_file\n",
    "from sqlalchemy import create_engine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The Genome Properties is a flat-file database that can be fount on Github.\n",
    "# The latest release of the database can be found at the following URL.\n",
    "\n",
    "genome_properties_database_url = 'https://raw.githubusercontent.com/ebi-pf-team/genome-properties/master/flatfiles/genomeProperties.txt'\n",
    "\n",
    "# For this tutorial, we will stream the file directly into the Jupyter notebook. Alternatively, \n",
    "# one could be downloaded the file with the UNIX wget or curl commands.\n",
    "\n",
    "with requests.Session() as current_download:\n",
    "    response = current_download.get(genome_properties_database_url, stream=True)\n",
    "    tree = parse_genome_properties_flat_file(StringIO(response.text))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1286"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# There are 1286 properties in the Genome Properties tree.\n",
    "len(tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coenzyme F420 utilization\n",
      "CRISPR region\n",
      "Reduction of oxidized methionine\n",
      "Phage: major features\n",
      "Resistance to Reactive Oxygen Species (ROS)\n",
      "tRNA aminoacylation\n",
      "Toxin-antitoxin system, type II\n",
      "Protein-coding palindromic elements\n",
      "Flagellar components of unknown function\n",
      "Bacillithiol utilization\n",
      "Toxin-antitoxin system, type I\n",
      "Toxin-antitoxin system, type III\n",
      "Abortive infection proteins\n",
      "Energy-coupling factor transporters\n",
      "Initiator caspases of the apoptosis extrinsic pathway\n",
      "Executor caspases of apoptosis\n"
     ]
    }
   ],
   "source": [
    "# Find all properties of type \"GUILD\".\n",
    "for genome_property in tree:\n",
    "    if genome_property.type == 'GUILD':\n",
    "        print(genome_property.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get property by identifier\n",
    "virulence = tree['GenProp0074']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GenProp0074, Type: CATEGORY, Name: Virulence, Thresh: 0, References: False, Databases: False, Steps: True, Parents: True, Children: True, Public: True"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "virulence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate to get the identifiers of child properties of virulence\n",
    "types_of_vir = [genprop.id for genprop in virulence.children]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "steps_of_type_3_secretion = [step.name for step in virulence.children[0].steps]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Type III secretion protein HpaP',\n",
       " 'Type III secretion system, HrpB1/HrpK',\n",
       " 'Type III secretion protein HrpB2',\n",
       " 'Type III secretion protein HrpB4',\n",
       " 'Type III secretion protein HrpB7',\n",
       " 'Type III secretion regulator, YopN/LcrE/InvE/MxiC',\n",
       " 'Type III secretion protein, LcrG',\n",
       " 'Type III secretion system, low calcium response, chaperone LcrH/SycD',\n",
       " 'Type III secretion system regulator, LcrR',\n",
       " 'Type III secretion apparatus protein OrgA/MxiK',\n",
       " 'Type III secretion system, PrgH/EprH',\n",
       " 'Surface presentation of antigens protein SpaK',\n",
       " 'Secretion system effector C, SseC-like',\n",
       " 'Tir chaperone protein (CesT) family',\n",
       " 'Type III secretion system chaperone SycN',\n",
       " 'Type III secretion system effector delivery regulator TyeA-related',\n",
       " 'YopD-like',\n",
       " 'Type III secretion system needle length determinant',\n",
       " 'Type III secretion system, needle protein',\n",
       " 'Proximal regulatory components',\n",
       " 'NolW-like',\n",
       " 'Secreted proteins, effectors',\n",
       " 'Type III secretion system chaperone YscB',\n",
       " 'Type III secretion system outer membrane pore YscC/HrcC',\n",
       " 'Type III secretion system YscD/HrpQ',\n",
       " 'Type III secretion system, secretion protein E',\n",
       " 'Type III secretion system YscG',\n",
       " 'DNType III secretion system effector YopR',\n",
       " 'Type III secretion system, YscI/HrpB',\n",
       " 'Type III secretion system lipoprotein HrcJ/YscJ',\n",
       " 'Type III secretion system, secretion protein K',\n",
       " 'Type III secretion system HrpE/YscL',\n",
       " 'ATPase, type III secretion system, H+-transporting',\n",
       " 'Type III secretion sytem,YscO',\n",
       " 'Type III secretion system apparatus protein YscQ/HrcQ/SpaO',\n",
       " 'Yop virulence translocation protein R',\n",
       " 'Type III secretion protein HrpO',\n",
       " 'Type III secretion protein SpaR/YscT',\n",
       " 'Type III exporter system, secretion apparatus protein BsaZ',\n",
       " 'Type III secretion protein HrcV',\n",
       " 'Type III secretion system chaperone, YscW',\n",
       " 'Type III secretion system YscX']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "steps_of_type_3_secretion"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creation and use of GenomePropertiesResults objects\n",
    "GenomePropertiesResults are used to compare property and step assignments across organisms programmatically."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse InterProScan files\n",
    "with open('E_coli_K12.tsv') as ipr5_file_one:\n",
    "    assignment_cache_1 = parse_interproscan_file(ipr5_file_one)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('E_coli_O157_H7.tsv') as ipr5_file_two:\n",
    "    assignment_cache_2 = parse_interproscan_file(ipr5_file_two)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create results comparison object\n",
    "results = GenomePropertiesResults(assignment_cache_1, assignment_cache_2, properties_tree=tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['E_coli_K12', 'E_coli_O157_H7']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.sample_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E_coli_K12</th>\n",
       "      <th>E_coli_O157_H7</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Property_Identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>GenProp0001</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GenProp0002</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GenProp0007</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GenProp0010</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GenProp0011</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GenProp2095</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GenProp2096</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GenProp2097</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GenProp2098</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GenProp2099</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1286 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    E_coli_K12 E_coli_O157_H7\n",
       "Property_Identifier                          \n",
       "GenProp0001                YES            YES\n",
       "GenProp0002                 NO             NO\n",
       "GenProp0007                YES            YES\n",
       "GenProp0010                 NO             NO\n",
       "GenProp0011                 NO             NO\n",
       "...                        ...            ...\n",
       "GenProp2095                 NO             NO\n",
       "GenProp2096                 NO             NO\n",
       "GenProp2097                 NO             NO\n",
       "GenProp2098                 NO             NO\n",
       "GenProp2099                 NO             NO\n",
       "\n",
       "[1286 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The property results property is used to compare two property assignments across samples.\n",
    "results.property_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>E_coli_K12</th>\n",
       "      <th>E_coli_O157_H7</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Property_Identifier</th>\n",
       "      <th>Step_Number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td rowspan=\"5\" valign=\"top\">GenProp0001</td>\n",
       "      <td>1</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td rowspan=\"5\" valign=\"top\">GenProp2099</td>\n",
       "      <td>7</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6525 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                E_coli_K12 E_coli_O157_H7\n",
       "Property_Identifier Step_Number                          \n",
       "GenProp0001         1                  YES            YES\n",
       "                    2                  YES            YES\n",
       "                    3                  YES            YES\n",
       "                    4                  YES            YES\n",
       "                    5                  YES            YES\n",
       "...                                    ...            ...\n",
       "GenProp2099         7                   NO             NO\n",
       "                    8                   NO             NO\n",
       "                    9                   NO             NO\n",
       "                    10                  NO             NO\n",
       "                    11                  NO             NO\n",
       "\n",
       "[6525 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The step results property is used to compare two step assignments across samples.\n",
    "results.step_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get properties with differing assignments\n",
    "results.differing_property_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get property assignments for virulence properties\n",
    "results.get_results(*types_of_vir, steps=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get step assignments for virulence properties\n",
    "results.get_results(*types_of_vir, steps=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get counts of virulence properties assigned YES, NO, and PARTIAL per organism\n",
    "results.get_results_summary(*types_of_vir, steps=False, normalize=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get counts of virulence steps assigned YES, NO, and PARTIAL per organism\n",
    "results.get_results_summary(*types_of_vir, steps=True, normalize=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get percentages of virulence steps assigned YES, NO, and PARTIAL per organism\n",
    "results.get_results_summary(*types_of_vir, steps=True, normalize=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creation and use of GenomePropertiesResultsWithMatches objects\n",
    "GenomePropertiesResultsWithMatches are an extension of GenomePropertiesResults objects that provide additional methods for comparing the InterProScan match information and protein sequences that support the existence of property steps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse InterProScan files and FASTA files\n",
    "with open('./E_coli_K12.tsv') as ipr5_file_one:\n",
    "    with open('./E_coli_K12.faa') as fasta_file_one:\n",
    "        extended_cache_one = parse_interproscan_file_and_fasta_file(ipr5_file_one, fasta_file_one)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse InterProScan files and FASTA files\n",
    "with open('./E_coli_O157_H7.tsv') as ipr5_file_two:\n",
    "    with open('./E_coli_O157_H7.faa') as fasta_file_two:\n",
    "        extended_cache_two = parse_interproscan_file_and_fasta_file(ipr5_file_two, fasta_file_two)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "extended_results = GenomePropertiesResultsWithMatches(extended_cache_one,\n",
    "                                                      extended_cache_two,\n",
    "                                                      properties_tree=tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GenomePropertiesResultsWithMatches objects possess the same \n",
    "# results comparison methods as GenomePropertiesResults objects\n",
    "extended_results.property_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "extended_results.step_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get matches and protein sequences that support properties and steps. \n",
    "extended_results.step_matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get only matches for K12\n",
    "extended_results.get_sample_matches('E_coli_K12', top=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "type_three_secretion_property_id = types_of_vir[0] # From section above.\n",
    "\n",
    "\n",
    "# Get matches for each Type III Secretion System component across both organisms.\n",
    "extended_results.get_property_matches(type_three_secretion_property_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get lowest E-value matches for each Type III Secretion System component for E_coli_O157_H7.\n",
    "extended_results.get_property_matches(type_three_secretion_property_id, sample='E_coli_O157_H7', top=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get lowest E-value matches for step 22 of Type III Secretion across both organisms. \n",
    "extended_results.get_step_matches(type_three_secretion_property_id, 22, top=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all matches for step 22 of Type III Secretion for E. coli K12. \n",
    "extended_results.get_step_matches(type_three_secretion_property_id, 22, top=False, sample='E_coli_K12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get skbio protein objects for a particular step.\n",
    "extended_results.get_supporting_proteins_for_step(type_three_secretion_property_id, 22, top=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### A note on Scikit-Bio\n",
    "Scikit-Bio is a numpy-based bioinformatics library that is a competitor to BioPython. Because it is numpy-based, it is quite fast and can be used to perform operations such as generating alignments and or producing phylogenetic trees. Pygenprop integrates Scikit-Bio for reading and writing FASTA files and the get_supporting_proteins_for_step() function of GenomePropertiesResultsWithMatches objects returns a list of Scikit-Bio Sequence objects. These Sequence objects can be aligned using Scikit-Bio and later incorporated into phylogenetic trees that compare the proteins that support a pathway step. Alignment and tree construction can be performed inside a Jupyter Notebook.  \n",
    "\n",
    "See the following documentation and tutorials for more information:\n",
    "\n",
    "- http://scikit-bio.org/docs/0.5.5/alignment.html\n",
    "- http://scikit-bio.org/docs/0.5.5/tree.html\n",
    "- https://nbviewer.jupyter.org/github/biocore/scikit-bio-cookbook/blob/master/Progressive%20multiple%20sequence%20alignment.ipynb\n",
    "- https://nbviewer.jupyter.org/github/biocore/scikit-bio-cookbook/blob/master/Alignments%20and%20phylogenetic%20reconstruction.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write FASTA file containing the sequences of the lowest E-value matches for \n",
    "# Type III Secretion System component 22 across both organisms.\n",
    "with open('type_3_step_22_top.faa', 'w') as out_put_fasta_file:\n",
    "    extended_results.write_supporting_proteins_for_step_fasta(out_put_fasta_file, \n",
    "                                                              type_three_secretion_property_id, \n",
    "                                                              22, top=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write FASTA file containing the sequences all matches for \n",
    "# Type III Secretion System component 22 across both organisms.\n",
    "with open('type_3_step_22_all.faa', 'w') as out_put_fasta_file:\n",
    "    extended_results.write_supporting_proteins_for_step_fasta(out_put_fasta_file, \n",
    "                                                              type_three_secretion_property_id, \n",
    "                                                              22, top=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading and writing Micromeda files\n",
    "Micromeda files are a new SQLite3-based pathway annotation storage format that allows for the simultaneous transfer of multiple organism's Genome Properties assignments and supporting information such as InterProScan annotations and protein sequences. These files allow for the transfer of complete Genome properties Datasets between researchers and software applications. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a SQLAlchemy engine object for a SQLite3 Micromeda file.  \n",
    "engine_no_proteins = create_engine('sqlite:///ecoli_compare_no_proteins.micro')\n",
    "\n",
    "# Write the results to the file.\n",
    "results.to_assignment_database(engine_no_proteins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a SQLAlchemy engine object for a SQLite3 Micromeda file.  \n",
    "engine_proteins = create_engine('sqlite:///ecoli_compare.micro')\n",
    "\n",
    "# Write the results to the file.\n",
    "extended_results.to_assignment_database(engine_proteins)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### A note on SQLAlchemy\n",
    "Because Pygenprop uses SQLAlchemy to write Micromeda files (SQlite3), it can also write assignment results and supporting information to a variety of relational databases.\n",
    "\n",
    "For example:\n",
    "\n",
    "```python\n",
    "create_engine('postgresql://scott:tiger@localhost/mydatabase')\n",
    "```\n",
    "\n",
    "See the following documentation for more information:\n",
    "\n",
    "- https://docs.sqlalchemy.org/en/13/core/engines.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load results from a Micromeda file.\n",
    "assignment_caches = load_assignment_caches_from_database(engine_no_proteins)\n",
    "results_reconstituted = GenomePropertiesResults(*assignment_caches, properties_tree=tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load results from a Micromeda file with proteins sequences.\n",
    "assignment_caches_with_proteins = load_assignment_caches_from_database_with_matches(engine_proteins)\n",
    "results_reconstituted_with_proteins = GenomePropertiesResultsWithMatches(*assignment_caches_with_proteins, \n",
    "                                                                         properties_tree=tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pygenprop CLI interface\n",
    "Pygenprop also includes a command line interface. A command-line tutorial can be found here [here](https://github.com/Micromeda/pygenprop/blob/improve-documentation/README.md#example-workflow)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "metadata": {
     "collapsed": false
    },
    "source": []
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
